python/common/org/python/types/Str.java

```diff
@@ -66,14 +66,76 @@ public Str(org.python.Object[] args, java.util.Map<java.lang.String, org.python.
     //     throw new org.python.exceptions.NotImplementedError("__init__() has not been implemented.");
     // }
 
+    private static boolean isCharPrintable(char c) {
+        // ASCII non-printable
+        if ((int) c <= 0x1f || (int) c >= 0x7f && (int) c <= 0xa0 || (int) c == 0xad) {
+            return false;
+        }
+        if ((int) c == 0x2029) {
+            return false;
+        }
+        if (Character.isISOControl(c)) {
+            return false;
+        }
+        return true;
+    }
+
     @org.python.Method(
             __doc__ = ""
     )
     public org.python.Object __repr__() {
-        String repr = this.value.replaceAll("\\n", "\\\\n").replaceAll("\\r", "\\\\r");
-        return new org.python.types.Str("'" + repr + "'");
+    /*
+    * Reference: https://www.python.org/dev/peps/pep-3138/#id7
+    * TODO: Need to treat the leading surrogate pair characters
+    */
+        StringBuilder sb = new StringBuilder();
+        boolean has_double_quote = false;
+        boolean has_single_quote = false;
+
+        for (char c : this.value.toCharArray()) {
+            if (c == '\'') {
+                has_single_quote = true;
+            } else if (c == '"') {
+                has_double_quote = true;
+            }
+
+            if (c == '\n') {
+                sb.append("\\n");
+            } else if (c == '\t') {
+                sb.append("\\t");
+            } else if (c == '\r') {
+                sb.append("\\r");
+            } else if (c == '\\') {
+                sb.append("\\\\");
+            // ASCII Non-Printable
+            } else if (c <= 0x1f || c >= 0x7f && c <= 0xa0 || c == 0xad) {
+                sb.append(String.format("\\x%02x", (int) c));
+            } else if (!this.isCharPrintable(c)) {
+                sb.append(String.format("\\u%04x", (int) c));
+            } else {
+                sb.append((char) c);
+            }
+        }
+
+        // Decide if we wanna wrap the result with single or double quotes
+        String quote;
+        String repr = sb.toString();
+
+        if (has_single_quote) {
+            if (has_double_quote) {
+                quote = new String("'");
```

Comment:
We can just have the string literal `"'"` here, instead of the `new String()` constructor – similarly for next two instances of this.